### PR TITLE
support append commit 

### PR DESCRIPTION
**Pull request by LiWeiJie**
[rust] allow append commit in dataset
[python] fix the missing mode in the dataset commit interface

### CODE REVIEW

File: rust/src/dataset.rs [hunk 1]
```diff
@@ -443,13 +443,15 @@ impl Dataset {
...
         let mut version = 1;
+        let mut dataset_fragments: Vec<Fragment> = Vec::new();
         let schema = if object_store.exists(&latest_manifest).await? {
             let dataset = Self::open(base_uri).await?;
             version = dataset.version().version + 1;
 
             if matches!(mode, WriteMode::Append) {
                 // Append mode: inherit indices from previous version.
                 indices = dataset.load_indices().await?;
+                dataset_fragments = dataset.fragments().iter().map(|f| f.clone()).collect();
```
Comment (wjones127) [inline rust/src/dataset.rs:454]: This will create duplicate paths in the current write append code path. The old fragments are already added here:

https://github.com/lancedb/lance/blob/d3e81539ee90a8c73d5c171a400d22248311737f/rust/src/dataset.rs#L355-L358

This is a breaking change in behavior of a public API, so I think I'd rather not make this change here if we can avoid it. But if we do make this change, we need to adjust the other code path as well, and add tests to make sure we aren't creating duplicate fragment entries.

File: python/python/lance/dataset.py [hunk 2]
```diff
@@ -603,7 +603,7 @@ def _commit(
             base_uri = str(base_uri)
         if not isinstance(new_schema, pa.Schema):
             raise TypeError(f"schema must be pyarrow.Schema, got {type(new_schema)}")
-        _Dataset.commit(base_uri, new_schema, fragments)
+        _Dataset.commit(base_uri, new_schema, fragments, mode)
```
Comment (wjones127) [inline python/python/lance/dataset.py:606]: +1 to this change.

File: rust/src/dataset.rs [hunk 3]
```diff
@@ -1119,6 +1123,41 @@ mod tests {
...
+            .await
+            .unwrap();
+        assert_eq!(20, dataset.count_rows().await.unwrap());
+
+        let fragments: Vec<Fragment> = dataset.fragments().iter().map(|f| f.clone()).collect();
+
+        let new_dataset =
+            Dataset::commit(test_uri, dataset.schema(), &fragments, WriteMode::Append)
+                .await
+                .unwrap();
```
Comment (wjones127) [inline rust/src/dataset.rs:1156]: I don't think this should be allowed because the added fragments have the same ids as the old ones.

**Review by wjones127 — CHANGES_REQUESTED** [review]
It seems like we need better validation when we making these commits. We can't have duplicate fragment ids, because that breaks the unique row ids.

**Comment by wjones127** [comment]
Perhaps what we want instead is a new struct to represent uncommitted fragments that doesn't have the ID

```rust
struct NewFragment {
    /// Files within the fragment.
    pub files: Vec<DataFile>,

    /// Optional file with deleted row ids.
    pub deletion_file: Option<DeletionFile>,
}
```

And then some function to append those to the log, adding new ids as appropriate:

```rust
impl Dataset {
    async fn append_new_fragments(dataset_uri: &str, new_fragments: &[NewFragment]) -> Result<Self> {
...}
}
```

That way we can handle the assignment of new fragment ids inside the function.

@eddyxu what do you think of that?

**Comment by wjones127** [comment]
@LiWeiJie closing this since we've implemented support in `_commit` for Append and several other operations in #1193

You can see an example of append here: https://github.com/lancedb/lance/blob/2ee7f01e739a56ca9de5f41aae5d6a9d8b8fc8a5/python/python/tests/test_dataset.py#L399-L421
